moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp

```diff
@@ -86,46 +86,60 @@ CostFn getCostFunctionFromStateValidator(const StateValidatorFn& state_validator
+      // The penalty of the validation function is added to the cost of the current timestep
+      // A state is rendered invalid if a cost results from the current validity check or if a penalty is carried over
+      // from the previous iteration
       Eigen::VectorXd current = values.col(timestep);
-      Eigen::VectorXd next = values.col(timestep + 1);
-      const double segment_distance = (next - current).norm();
-      double interpolation_fraction = 0.0;
-      const double interpolation_step = std::min(0.5, interpolation_step_size / segment_distance);
-      bool found_invalid_state = false;
-      double penalty = 0.0;
-      while (!found_invalid_state && interpolation_fraction < 1.0)
+      costs(timestep) += state_validator_fn(current);
+      bool found_invalid_state = costs(timestep) > 0.0;
+
+      // If state is valid, interpolate towards the next waypoint if there is one
+      bool continue_interpolation =
+          !found_invalid_state && timestep < (values.cols() - 1) && interpolation_step_size > 0.0;
+      if (continue_interpolation)
       {
-        Eigen::VectorXd sample_vec = (1 - interpolation_fraction) * current + interpolation_fraction * next;
+        Eigen::VectorXd next = values.col(timestep + 1);
+        const double interpolation_step = std::min(0.5, interpolation_step_size / (next - current).norm());
+        for (double interpolation_fraction = interpolation_step; interpolation_fraction < 1.0;
+             interpolation_fraction += interpolation_step)
+        {
+          Eigen::VectorXd sample_vec = (1 - interpolation_fraction) * current + interpolation_fraction * next;
 
-        penalty = state_validator_fn(sample_vec);
-        found_invalid_state = penalty > 0.0;
-        interpolation_fraction += interpolation_step;
+          double penalty = state_validator_fn(sample_vec);
+          found_invalid_state = penalty > 0.0;
+          if (found_invalid_state)
+          {
+            // Apply weighted penalties -> This trajectory is definitely invalid
+            costs(timestep) += (1 - interpolation_fraction) * penalty;
+            costs(timestep + 1) += interpolation_fraction * penalty;
+            break;
+          }
+        }
       }
 
+      // Track groups of invalid states as "invalid windows" for subsequent smoothing
       if (found_invalid_state)
       {
-        // Apply weighted penalties -> This trajectory is definitely invalid
-        costs(timestep) = (1 - interpolation_fraction) * penalty;
-        costs(timestep + 1) = interpolation_fraction * penalty;
+        // Mark solution as invalid
         validity = false;
 
-        // Open new invalid window when this is the first detected invalid state in a group
+        // OPEN new invalid window when this is the first detected invalid state in a group
         if (!in_invalid_window)
         {
-          invalid_windows.emplace_back(timestep, values.cols());
+          invalid_windows.emplace_back(timestep, timestep);
```

Comment:
Why do we add current state to the current state?

Reply:
Can you add a descriptive comment for `invalid_windows` when it is initialized earlier in the code?

Reply:
What is missing from the description at the top of the loop?

Reply:
I added more remarks to the documentation